File: cpp/code_forces/edu/itmo/seg_tree1/part4/1.py
class SegmentTree:
    def __init__(self, arr):
        self.n = len(arr)
        self.tree = [0] * (4 * self.n)
        self.build(arr, 1, 0, self.n - 1)

    def build(self, arr, v, tl, tr):
        if tl == tr:
            self.tree[v] = arr[tl] if tl % 2 == 0 else -arr[tl]
        else:
            tm = (tl + tr) // 2
            self.build(arr, 2 * v, tl, tm)
            self.build(arr, 2 * v + 1, tm + 1, tr)
            self.tree[v] = self.tree[2 * v] + self.tree[2 * v + 1]

    def update(self, v, tl, tr, pos, new_val):
        if tl == tr:
            self.tree[v] = new_val if tl % 2 == 0 else -new_val
        else:
            tm = (tl + tr) // 2
            if pos <= tm:
                self.update(2 * v, tl, tm, pos, new_val)
            else:
                self.update(2 * v + 1, tm + 1, tr, pos, new_val)
            self.tree[v] = self.tree[2 * v] + self.tree[2 * v + 1]

    def query(self, v, tl, tr, l, r):
        if l > r:
            return 0
        if l == tl and r == tr:
            return self.tree[v]
        tm = (tl + tr) // 2
        return self.query(2 * v, tl, tm, l, min(r, tm)) + self.query(2 * v + 1, tm + 1, tr, max(l, tm + 1), r)

def alternating_sum(arr, queries):
    seg_tree = SegmentTree(arr)
    result = []
    for query in queries:
        if query[0] == 0:
            seg_tree.update(1, 0, seg_tree.n - 1, query[1] - 1, query[2])
        else:
            l, r = query[1], query[2]
            result.append(seg_tree.query(1, 0, seg_tree.n - 1, l - 1, r - 1) * (-1 if l%2==0 else 1))
    return result

# Чтение входных данных
n = int(input())
arr = list(map(int, input().split()))
m = int(input())
queries = []
for _ in range(m):
    query = list(map(int, input().split()))
    queries.append(query)

# Получение и вывод результатов
result = alternating_sum(arr, queries)
for res in result:
    print(res)

File: cpp/code_forces/edu/itmo/seg_tree1/part4/1.cpp
#include <bits/stdc++.h>
using namespace std;

struct segtree
{
    vector<long long> tree;
    int size;

    segtree(vector<int> &arr)
    {
        size = 1;
        while (size < arr.size())
            size *= 2;
        tree.assign(size * 2 - 1, 0);
        build(arr, 0, 0, size);
    }

    void build(const vector<int> &arr, int x, int lx, int rx)
    {
        if (rx - lx == 1)
        {
            if (lx < arr.size())
            {
                tree[x] = lx % 2 == 1 ? arr[lx] : -arr[lx];
            }
            return;
        }
        int m = (lx + rx) / 2;
        build(arr, x * 2 + 1, lx, m);
        build(arr, x * 2 + 2, m, rx);
        tree[x] = tree[x * 2 + 1] + tree[x * 2 + 2];
    }

    void set(int i, int v)
    {
        set(i, v, 0, 0, size);
    }

    void set(int i, int v, int x, int lx, int rx)
    {
        if (rx - lx == 1)
        {
            tree[x] = lx % 2 == 1 ? v : -v;
            return;
        }
        int m = (lx + rx) / 2;
        if (i < m)
        {
            set(i, v, x * 2 + 1, lx, m);
        }
        else
        {
            set(i, v, x * 2 + 2, m, rx);
        }
        tree[x] = tree[x * 2 + 1] + tree[x * 2 + 2];
    }

    long long sum(int l, int r)
    {
        return sum(l, r, 0, 0, size);
    }

    long long sum(int l, int r, int x, int lx, int rx)
    {
        if (r <= lx || l >= rx)
        {
            return 0;
        }
        if (l <= lx && r >= rx)
        {
            return tree[x];
        }
        int m = (lx + rx) / 2;
        return sum(l, r, x * 2 + 1, lx, m) + sum(l, r, x * 2 + 2, m, rx);
    }
};

int main()
{
    int n, m;
    cin >> n;
    vector<int> v(n);
    for (auto &x : v)
    {
        cin >> x;
    }
    segtree t(v);
    cin >> m;
    while (m-- > 0)
    {
        int q, l, r;
        cin >> q;
        if (!q)
        {
            cin >> l >> r;
            t.set(l - 1, r);
        }
        else
        {
            cin >> l >> r;
            cout << t.sum(l - 1, r) * (l % 2 == 0 ? 1 : -1) << "\n";
        }
    }
    cout << endl;
    return 0;
}